tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java

```diff
@@ -18,19 +22,29 @@
   @Autowired
   private ApplicationContext applicationContext;
 
+  @Autowired
+  private TasksProperties tasksProperties;
+
   private volatile List<ITaskHandler> handlers;
 
   @Override
   public ITaskHandler getTaskHandler(IBaseTask task) {
-    List<ITaskHandler> results = getHandlers().stream().filter(h -> h.handles(task)).collect(Collectors.toList());
-
-    if (results.size() > 1) {
-      String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
-      throw new IllegalStateException("Too many handlers are able to handle task of type '" + task.getType() + "': " + resultsSt);
-    } else if (results.size() == 1) {
-      return results.get(0);
+    // TODO: Should we add caching here for larger applications?
```

Comment:
Theoretically the application could have some state dependent implementations, probably not the case though.